File: answers/other/answers/leetcode/Lowest Common Ancestor of a Binary Search Tree.cpp
//@type Tree
//@result 27 / 27 test cases passed. Status: Accepted Runtime: 40 ms Submitted: 1 minute ago You are here! Your runtime beats 43.06% of cpp submissions.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (NULL == root || root == p || root == q) {
			return root;
		}
		else {
			TreeNode *left = lowestCommonAncestor(root->left, p, q);
			TreeNode *right = lowestCommonAncestor(root->right, p, q);
			return NULL == left ? right : (NULL == right ? left : root);
		}
    }
};

File: answers/leetcode/Lowest Common Ancestor of a Binary Search Tree/Lowest Common Ancestor of a Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return null;
        }
        int pDiff = p.val - root.val;
        int qDiff = q.val - root.val;
        if (pDiff * qDiff <= 0) {
            return root;
        } else if (pDiff > 0 && qDiff > 0) {
            return this.lowestCommonAncestor(root.right, p, q);
        } else {
            return this.lowestCommonAncestor(root.left, p, q);
        }
    }
}